rfcs/0006-transaction.md

```diff
@@ -0,0 +1,366 @@
+
+This approach can be adjusted to support SI as well by only tracking the keys that are being written to. When user set the isolation level to SI, the `read_keys` could always be set as empty. The GC mechanism to release inactive transactions is still the same as the SSI.
+
+#### GC & Expiration
+
+As described above, the entries in `recent_committed_txns` could be released whenever ALL the running transactions' sequence number are bigger than the committed sequence number of the entry. 
+
+We have to track the running transactions's sequence number in the `Oracle`.
+
+Also, it's less make sense to allow a transaction to be running for 12 hours, the `recent_committed_txns` will also keep growing for such 12 hours, and the GC will not be able to release the entries. This is a potential problem that the memory usage of the `Oracle` will keep growing, and possible lead to OOM.
+
+To mitigate this, we should allow users to set a default expiration time for all of the transactions & snapshots, and the transactions & snapshots will be considered as expired if it's not finished within the expiration time.
+
+```rust
+struct Oracle {
+    running_txns: HashSet<u64>,
+    next_seq: u64,
+    discard_seq: u64,
+    recent_committed: Deque<TransactionState>>,
+}
+
+struct TransactionState {
+    started_time: u64,
+    expired_time: u64,
+    started_seq: u64,
+    write_keys: HashSet<Bytes>,
+    read_keys: Vec<Range<Bytes>>,
+    committed_seq: Option<u64>
+}
+```
+
+We should let the `Oracle` to track the running transactions in `running_txns`, and when the transaction is finished, whether it's committed or rolled back, the `Oracle` should remove the transaction from the `running_txns` set. If the transaction is expired, the `Oracle` should also remove the transaction from the `running_txns` set.
+
+Please note that Snapshot could also be considered as a read only transaction, and it should also be tracked in `running_txns`. And Snapshot should also has a default expiration as well.
+
+With `running_txns`, we could find out the earliest running transaction's sequence number, and `Oracle` could use this sequence number to determine which entries in `recent_committed_txns` could be GCed. This could help us to ensure both `running_txns` and `recent_committed_txns` do not grow out of a time limit.
+
+Also, we can maintain a `discard_seq` during `recent_committed_txns`'s cleanup. If the cleaned up txn's create time is earlier than global expiration time, we could set `discard_seq` to the txn's sequence number, and the compactor could respect this `discard_seq` to discard the values earlier than `discard_seq` safely when it found a value with multiple versions.
```

Comment:
can you share more on this in the comment? I don't understand the relation between this and compaction.

Reply:
https://github.com/google/leveldb/blob/23e35d792b9154f922b8b575b12596a4d8664c65/db/db_impl.cc#L966

it's similiar with leveldb's `smallest_snapshot`, when a kv have multiple versions, the compactor should reclaime as many rows as possible, but it can not reclaime the rows whose seqnum bigger than `smallest_snapshot`

Reply:
thank you for the advice, let me think about this. i'm planned to add an RFC about the cleanup logic of Snapshot after https://github.com/slatedb/slatedb/pull/389 get merged.

the creation & use of Snapshot will be very trivial after seqnum got ready in everywhere, the only thing left we need carefully consider is the cleanup part.

i'll put your advice as one of the choices in this RFC.